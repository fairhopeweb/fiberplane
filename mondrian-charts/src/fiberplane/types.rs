use crate::types::*;
pub use fiberplane_models::notebooks::{GraphType, StackingType};
pub use fiberplane_models::providers::{Metric, ProviderEvent, Timeseries};
pub use fiberplane_models::timestamps::{TimeRange, Timestamp};
use serde::Serialize;
use std::collections::BTreeMap;

/// All the data necessary to generate an abstract chart from an array of
/// timeseries.
pub struct TimeseriesSourceData<'source, 'slice, 'value> {
    /// The type of chart to display.
    pub graph_type: GraphType,

    /// The type of stacking to apply to the chart.
    pub stacking_type: StackingType,

    /// Array of timeseries data to display in the chart.
    ///
    /// Make sure the timeseries contains data for the given time range, or you
    /// may not see any results.
    pub timeseries_data: &'slice [&'source Timeseries],

    /// The time range to be displayed.
    pub time_range: TimeRange,

    /// Additional values that will be plotted on the chart, but which are not
    /// part of any timeseries. These are not plotted, but are taken into
    /// account when deciding the range of the Y axis.
    pub additional_values: &'value [f64],
}

/// All the data necessary to generate an abstract chart from a combination of
/// timeseries data, events and an optional target latency.
pub struct CombinedSourceData<'source, 'slice> {
    /// The type of stacking to apply to the chart.
    ///
    /// **Warning:** This property is accepted for consistency, but setting it
    /// to anything except [GraphType::Line] will cause the events to be
    /// ignored.
    pub graph_type: GraphType,

    /// The type of stacking to apply to the chart.
    pub stacking_type: StackingType,

    /// Array of timeseries data to display in the chart.
    ///
    /// Make sure the timeseries contains data for the given time range, or you
    /// may not see any results.
    pub timeseries_data: &'slice [&'source Timeseries],

    /// Array of events to display in the chart.
    ///
    /// Note that events will not be displayed if the `graph_type` is anything
    /// other than [`GraphType::Line`].
    pub events: &'slice [&'source ProviderEvent],

    /// Optional target latency to display on the chart, in seconds.
    pub target_latency: Option<f64>,

    /// The time range to be displayed.
    pub time_range: TimeRange,
}

/// Source type for series in charts that contain combined data sources.
#[derive(Serialize)]
#[non_exhaustive]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SeriesSource<'source> {
    Timeseries(&'source Timeseries),
    Events,
    TargetLatency,
}

impl<'source> From<&'source Timeseries> for SeriesSource<'source> {
    fn from(value: &'source Timeseries) -> Self {
        Self::Timeseries(value)
    }
}

/// Source type for points in charts that contain combined data sources.
#[derive(Serialize)]
#[non_exhaustive]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PointSource<'source> {
    Metric(&'source Metric),
    Event(&'source ProviderEvent),
    None,
}

impl<'source> From<&'source Metric> for PointSource<'source> {
    fn from(value: &'source Metric) -> Self {
        Self::Metric(value)
    }
}

impl<'source> From<MondrianChart<&'source Timeseries, &'source Metric>>
    for MondrianChart<SeriesSource<'source>, PointSource<'source>>
{
    fn from(value: MondrianChart<&'source Timeseries, &'source Metric>) -> Self {
        Self {
            shape_lists: value.shape_lists.into_iter().map(ShapeList::into).collect(),
            x_axis: value.x_axis,
            y_axis: value.y_axis,
        }
    }
}

/// A map that holds arbitrary values per timestamp.
///
/// The keys in the map represent timestamps, while the values are generic and
/// are calculated over all the metrics matching that timestamp.
pub(crate) type Buckets<T> = BTreeMap<Timestamp, T>;

/// Represents the minimum and maximum values inside a series of numbers.
#[derive(Clone, Copy)]
pub(crate) struct MinMax(pub f64, pub f64);

impl MinMax {
    pub fn from_value(value: f64) -> Self {
        Self(value, value)
    }

    #[must_use]
    pub fn extend_with_other(mut self, other: Self) -> Self {
        if other.0 < self.0 {
            self.0 = other.0;
        }
        if other.1 > self.1 {
            self.1 = other.1;
        }

        self
    }

    #[must_use]
    pub fn extend_with_value(mut self, value: f64) -> Self {
        if value < self.0 {
            self.0 = value;
        } else if value > self.1 {
            self.1 = value;
        }

        self
    }
}

impl<'source> From<ShapeList<&'source Timeseries, &'source Metric>>
    for ShapeList<SeriesSource<'source>, PointSource<'source>>
{
    fn from(value: ShapeList<&'source Timeseries, &'source Metric>) -> Self {
        Self {
            shapes: value.shapes.into_iter().map(Shape::into).collect(),
            source: value.source.into(),
        }
    }
}

impl<'source> From<Shape<&'source Metric>> for Shape<PointSource<'source>> {
    fn from(other: Shape<&'source Metric>) -> Self {
        match other {
            Shape::Area(area) => Shape::Area(area.into()),
            Shape::Line(line) => Shape::Line(line.into()),
            Shape::Point(point) => Shape::Point(point.into()),
            Shape::Rectangle(rectangle) => Shape::Rectangle(rectangle.into()),
        }
    }
}

impl<'source> From<Area<&'source Metric>> for Area<PointSource<'source>> {
    fn from(value: Area<&'source Metric>) -> Self {
        Self {
            area_gradient_shown: None,
            points: value.points.into_iter().map(AreaPoint::into).collect(),
        }
    }
}

impl<'source> From<AreaPoint<&'source Metric>> for AreaPoint<PointSource<'source>> {
    fn from(value: AreaPoint<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y_min: value.y_min,
            y_max: value.y_max,
            source: value.source.into(),
        }
    }
}

impl<'source> From<Line<&'source Metric>> for Line<PointSource<'source>> {
    fn from(value: Line<&'source Metric>) -> Self {
        Self {
            area_gradient_shown: None,
            points: value.points.into_iter().map(Point::into).collect(),
        }
    }
}

impl<'source> From<Point<&'source Metric>> for Point<PointSource<'source>> {
    fn from(value: Point<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y: value.y,
            source: value.source.into(),
        }
    }
}

impl<'source> From<Rectangle<&'source Metric>> for Rectangle<PointSource<'source>> {
    fn from(value: Rectangle<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y: value.y,
            width: value.width,
            height: value.height,
            source: value.source.into(),
        }
    }
}

pub(crate) type StackedChartBuckets = Buckets<StackedChartBucketValue>;

#[derive(Clone, Copy, Default)]
pub(crate) struct StackedChartBucketValue {
    /// Used to keep track of how much a bucket is "filled" while calculating
    /// the area shapes.
    pub current_y: f64,

    /// The sum of all values in the bucket.
    pub total: f64,
}
