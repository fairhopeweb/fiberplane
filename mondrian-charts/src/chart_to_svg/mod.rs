mod constants;
mod create_area_path_def;
mod create_line_path_def;
mod generate_chart_content_svg;
mod generate_grid_with_axes_svg;
mod scales;
mod tick_formatters;

use crate::MondrianChart;
use constants::*;
use generate_chart_content_svg::generate_chart_content_svg;
use generate_grid_with_axes_svg::generate_grid_with_axes_svg;
use scales::Scales;
use std::borrow::Cow;

pub use self::tick_formatters::FormatterKind;

pub struct ChartOptions<'a, S> {
    pub width: u16,
    pub height: u16,

    pub area_gradient_shown: bool,
    pub grid_borders_shown: bool,
    pub grid_shown: bool,
    pub grid_stroke_color: String,
    pub grid_stroke_dasharray: Vec<f32>,
    pub get_shape_list_color: &'a dyn Fn(&'a S, usize) -> &'a str,
    pub tick_color: String,
    pub x_formatter: FormatterKind,
    pub y_formatter: FormatterKind,
}

pub fn chart_to_svg<'a, S, P>(
    chart: &'a MondrianChart<S, P>,
    options: &ChartOptions<'a, S>,
) -> String {
    let ChartOptions { width, height, .. } = options;

    let x_max = width - MARGIN_LEFT - MARGIN_RIGHT;
    let y_max = height - MARGIN_TOP - MARGIN_BOTTOM;

    let scales = Scales::new(x_max as f64, y_max as f64);

    let clip_path_id = "clip-path";
    let clip_path_y_start = -(CHART_SHAPE_OVERFLOW_MARGIN as i32);
    let clip_path_height = y_max + 2 * CHART_SHAPE_OVERFLOW_MARGIN;

    let grid = if options.grid_shown {
        Cow::Owned(generate_grid_with_axes_svg(chart, &scales, options))
    } else {
        Cow::Borrowed("")
    };

    let chart_content = generate_chart_content_svg(chart, &scales, options);

    format!(
        "<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" width=\"{width}\" height=\"{height}\">\
            <defs>\
                <clipPath id=\"{clip_path_id}\">\
                  <rect x=\"0\" y=\"{clip_path_y_start}\" width=\"{x_max}\" height=\"{clip_path_height}\" />\
                </clipPath>\
            </defs>\
            <g transform=\"translate({MARGIN_LEFT}, {MARGIN_TOP})\">\
              {grid}\
              <g clip-path=\"url(#{clip_path_id})\">{chart_content}</g>\
            </g>\
        </svg>"
    )
}
