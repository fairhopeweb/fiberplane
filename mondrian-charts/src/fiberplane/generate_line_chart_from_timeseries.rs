use super::utils::*;
use crate::fiberplane::{Metric, MinMax, Timeseries, TimeseriesSourceData};
use crate::types::{Axis, Line, MondrianChart, Point, Shape, ShapeList};
use std::convert::identity;

pub(crate) fn generate_line_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let buckets = create_metric_buckets(input.timeseries_data, |min_max, value| {
        min_max
            .map(|min_map: MinMax| min_map.extend_with_value(value))
            .unwrap_or_else(|| MinMax::from_value(value))
    });

    let mut x_axis = get_x_axis_from_time_range(&input.time_range);
    let mut y_axis = calculate_y_axis_range(&buckets, identity);

    for &value in input.additional_values {
        y_axis = y_axis.extend_with_value(value);
    }

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                create_shapes(&timeseries.metrics, &x_axis, &y_axis, interval)
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

fn create_shapes<'source>(
    metrics: &'source [Metric],
    x_axis: &Axis,
    y_axis: &Axis,
    interval: Option<f64>,
) -> Vec<Shape<&'source Metric>> {
    match metrics.len() {
        0 => Vec::new(),
        1 => {
            let metric = &metrics[0];
            if metric.value.is_nan() {
                Vec::new()
            } else {
                vec![Shape::Point(create_point_for_metric(
                    metric, x_axis, y_axis,
                ))]
            }
        }
        _ => {
            split_into_continuous_lines(metrics, interval)
                .into_iter()
                .map(|line| {
                    // If the line only containes one metric value, render it as a point
                    // Otherwise, render a line
                    if line.len() == 1 {
                        Shape::Point(create_point_for_metric(line[0], x_axis, y_axis))
                    } else {
                        Shape::Line(Line {
                            area_gradient_shown: None,
                            points: line
                                .into_iter()
                                .map(|metric| create_point_for_metric(metric, x_axis, y_axis))
                                .collect(),
                        })
                    }
                })
                .collect()
        }
    }
}

fn create_point_for_metric<'source>(
    metric: &'source Metric,
    x_axis: &Axis,
    y_axis: &Axis,
) -> Point<&'source Metric> {
    let time = get_time_from_timestamp(metric.time);

    Point {
        x: normalize_along_linear_axis(time, x_axis),
        y: normalize_along_linear_axis(metric.value, y_axis),
        source: metric,
    }
}
