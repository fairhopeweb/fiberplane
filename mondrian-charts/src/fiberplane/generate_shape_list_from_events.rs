use super::utils::*;
use crate::fiberplane::{PointSource, ProviderEvent, SeriesSource};
use crate::types::{Area, AreaPoint, Axis, Shape, ShapeList};

pub(crate) fn generate_shape_list_from_events<'source>(
    x_axis: &Axis,
    events: &[&'source ProviderEvent],
) -> ShapeList<SeriesSource<'source>, PointSource<'source>> {
    ShapeList {
        shapes: events
            .iter()
            .map(|event| {
                Shape::Area(Area {
                    area_gradient_shown: None,
                    points: vec![AreaPoint {
                        x: normalize_along_linear_axis(get_time_from_timestamp(event.time), x_axis),
                        y_min: 0.,
                        y_max: 1.,
                        source: PointSource::Event(event),
                    }],
                })
            })
            .collect(),
        source: SeriesSource::Events,
    }
}
