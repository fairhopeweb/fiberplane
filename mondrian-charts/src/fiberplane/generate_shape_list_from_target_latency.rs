use super::utils::*;
use crate::fiberplane::{PointSource, SeriesSource};
use crate::types::{Axis, Shape, ShapeList};
use crate::{Line, Point};

/// Generates a shape list with a single line from a fixed target latency.
pub(crate) fn generate_shape_list_from_target_latency<'source>(
    y_axis: &Axis,
    value: f64,
) -> ShapeList<SeriesSource<'source>, PointSource<'source>> {
    ShapeList {
        shapes: vec![Shape::Line(Line {
            area_gradient_shown: Some(false),
            points: vec![
                Point {
                    x: 0.,
                    y: normalize_along_linear_axis(value, y_axis),
                    source: PointSource::None,
                },
                Point {
                    x: 1.,
                    y: normalize_along_linear_axis(value, y_axis),
                    source: PointSource::None,
                },
            ],
        })],
        source: SeriesSource::TargetLatency,
    }
}
