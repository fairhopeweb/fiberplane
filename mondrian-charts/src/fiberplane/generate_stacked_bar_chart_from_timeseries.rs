use super::utils::*;
use crate::fiberplane::{Metric, StackedChartBuckets, Timeseries, TimeseriesSourceData};
use crate::types::{Axis, MondrianChart, Rectangle, Shape, ShapeList};

pub(crate) fn generate_stacked_bar_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let BucketsAndAxes {
        mut buckets,
        is_percentage,
        mut x_axis,
        y_axis,
    } = calculate_buckets_and_axes_for_stacked_chart(&input);

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        x_axis = x_axis.extend_with_interval(interval);
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let mut bar_args = BarArgs {
        bar_width: calculate_bar_width(&x_axis, interval, 1),
        buckets: &mut buckets,
        is_percentage,
        x_axis: &x_axis,
        y_axis: &y_axis,
    };

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                timeseries
                    .metrics
                    .iter()
                    .filter_map(|metric| create_bar_shape(metric, &mut bar_args))
                    .collect()
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

struct BarArgs<'axes, 'buckets> {
    bar_width: f64,
    buckets: &'buckets mut StackedChartBuckets,
    is_percentage: bool,
    x_axis: &'axes Axis,
    y_axis: &'axes Axis,
}

fn create_bar_shape<'source>(
    metric: &'source Metric,
    args: &mut BarArgs,
) -> Option<Shape<&'source Metric>> {
    if metric.value.is_nan() {
        return None;
    };

    let Some(bucket_value) = args.buckets.get_mut(&metric.time) else {
        return None;
    };

    let time = get_time_from_timestamp(metric.time);
    let value = if args.is_percentage {
        metric.value / bucket_value.total
    } else {
        metric.value
    };

    let x = normalize_along_linear_axis(time, args.x_axis) - 0.5 * args.bar_width;
    let y = bucket_value.current_y;

    let height = normalize_along_linear_axis(value, args.y_axis);
    bucket_value.current_y += height;

    Some(Shape::Rectangle(Rectangle {
        x,
        y,
        width: args.bar_width,
        height,
        source: metric,
    }))
}
