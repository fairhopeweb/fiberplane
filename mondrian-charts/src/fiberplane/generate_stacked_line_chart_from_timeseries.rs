use super::utils::*;
use crate::fiberplane::{Metric, StackedChartBuckets, Timeseries, TimeseriesSourceData};
use crate::types::{Area, AreaPoint, Axis, MondrianChart, Shape, ShapeList};

pub(crate) fn generate_stacked_line_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let BucketsAndAxes {
        mut buckets,
        is_percentage,
        mut x_axis,
        mut y_axis,
    } = calculate_buckets_and_axes_for_stacked_chart(&input);

    for &value in input.additional_values {
        y_axis = y_axis.extend_with_value(value);
    }

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let mut args = BarArgs {
        buckets: &mut buckets,
        is_percentage,
        x_axis: &x_axis,
        y_axis: &y_axis,
    };

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                create_shapes(&timeseries.metrics, interval, &mut args)
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

struct BarArgs<'axes, 'buckets> {
    buckets: &'buckets mut StackedChartBuckets,
    is_percentage: bool,
    x_axis: &'axes Axis,
    y_axis: &'axes Axis,
}

fn create_shapes<'source>(
    metrics: &'source [Metric],
    interval: Option<f64>,
    args: &mut BarArgs,
) -> Vec<Shape<&'source Metric>> {
    split_into_continuous_lines(metrics, interval)
        .iter()
        .map(|line| {
            Shape::Area(Area {
                area_gradient_shown: None,
                points: line
                    .iter()
                    .filter_map(|metric| create_point_for_metric(metric, args))
                    .collect(),
            })
        })
        .collect()
}

fn create_point_for_metric<'source>(
    metric: &'source Metric,
    args: &mut BarArgs,
) -> Option<AreaPoint<&'source Metric>> {
    let bucket_value = args.buckets.get_mut(&metric.time)?;

    let time = get_time_from_timestamp(metric.time);
    let value = if args.is_percentage {
        metric.value / bucket_value.total
    } else {
        metric.value
    };

    let y_min = bucket_value.current_y;
    let y_max = y_min + normalize_along_linear_axis(value, args.y_axis);
    bucket_value.current_y = y_max;

    Some(AreaPoint {
        x: normalize_along_linear_axis(time, args.x_axis),
        y_min,
        y_max,
        source: metric,
    })
}
