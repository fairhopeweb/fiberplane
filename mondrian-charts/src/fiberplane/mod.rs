mod constants;
mod generate_bar_chart_from_timeseries;
mod generate_line_chart_from_timeseries;
mod generate_shape_list_from_events;
mod generate_shape_list_from_target_latency;
mod generate_stacked_bar_chart_from_timeseries;
mod generate_stacked_line_chart_from_timeseries;
mod types;
mod utils;

#[cfg(test)]
mod tests;

use generate_bar_chart_from_timeseries::generate_bar_chart_from_timeseries;
use generate_line_chart_from_timeseries::generate_line_chart_from_timeseries;
use generate_shape_list_from_events::generate_shape_list_from_events;
use generate_stacked_bar_chart_from_timeseries::generate_stacked_bar_chart_from_timeseries;
use generate_stacked_line_chart_from_timeseries::generate_stacked_line_chart_from_timeseries;
pub use types::*;

use crate::MondrianChart;

use self::generate_shape_list_from_target_latency::generate_shape_list_from_target_latency;

/// Generates an abstract chart from a combination of timeseries data, events
/// and an optional target latency.
///
/// May return `None` if the graph type is unrecognized.
pub fn generate<'source>(
    input: CombinedSourceData<'source, '_>,
) -> Option<MondrianChart<SeriesSource<'source>, PointSource<'source>>> {
    let CombinedSourceData {
        graph_type,
        stacking_type,
        timeseries_data,
        events,
        target_latency,
        time_range,
    } = input;

    let chart = generate_from_timeseries(TimeseriesSourceData {
        graph_type,
        stacking_type,
        timeseries_data,
        time_range,
        additional_values: &target_latency.iter().cloned().collect::<Vec<_>>(),
    })?;

    let mut chart: MondrianChart<SeriesSource, PointSource> = chart.into();

    if graph_type == GraphType::Line && !events.is_empty() {
        chart
            .shape_lists
            .push(generate_shape_list_from_events(&chart.x_axis, events));
    }

    if let Some(latency) = target_latency {
        chart
            .shape_lists
            .push(generate_shape_list_from_target_latency(
                &chart.y_axis,
                latency,
            ));
    }

    Some(chart)
}

/// Generates an abstract chart from the given timeseries data.
///
/// May return `None` if the graph type is unrecognized.
pub fn generate_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> Option<MondrianChart<&'source Timeseries, &'source Metric>> {
    let abstract_chart = match (input.graph_type, input.stacking_type) {
        (GraphType::Line, StackingType::None) => generate_line_chart_from_timeseries(input),
        (GraphType::Line, _) => generate_stacked_line_chart_from_timeseries(input),
        (GraphType::Bar, StackingType::None) => generate_bar_chart_from_timeseries(input),
        (GraphType::Bar, _) => generate_stacked_bar_chart_from_timeseries(input),
        (_, _) => return None,
    };

    Some(abstract_chart)
}
