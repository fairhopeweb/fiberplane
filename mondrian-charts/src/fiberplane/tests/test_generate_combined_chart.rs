use crate::chart_to_svg::*;
use crate::fiberplane::{
    generate, CombinedSourceData, GraphType, Metric, ProviderEvent, SeriesSource, StackingType,
    TimeRange, Timeseries, Timestamp,
};
use std::str::FromStr;

#[test]
fn test_generate_line_chart_from_timeseries() {
    let events = vec![
        get_event_at_minute(2, "deploy 1"),
        get_event_at_minute(7, "deploy 2"),
    ];

    let timeseries = Timeseries::builder()
        .name("dummy_data")
        .metrics(vec![
            get_metric_at_minute(0, 10.),
            get_metric_at_minute(1, 15.),
            get_metric_at_minute(2, 20.),
            get_metric_at_minute(4, 25.),
            get_metric_at_minute(5, 30.),
            get_metric_at_minute(6, 25.),
            get_metric_at_minute(7, 20.),
            get_metric_at_minute(8, 15.),
            get_metric_at_minute(9, 10.),
        ])
        .visible(true)
        .build();

    let target_latency = 40.;

    let chart = generate(CombinedSourceData {
        graph_type: GraphType::Line,
        stacking_type: StackingType::None,
        time_range: TimeRange {
            from: get_date_at_minute(0),
            to: get_date_at_minute(10),
        },
        timeseries_data: &[&timeseries],
        events: &events.iter().collect::<Vec<_>>(),
        target_latency: Some(target_latency),
    })
    .unwrap();

    insta::assert_yaml_snapshot!(chart);

    let svg = chart_to_svg(
        &chart,
        &ChartOptions {
            width: 640,
            height: 480,
            area_gradient_shown: true,
            grid_borders_shown: true,
            grid_shown: false,
            grid_stroke_color: "#e7e7e7".to_owned(),
            grid_stroke_dasharray: Default::default(),
            get_shape_list_color: &|source, _| match source {
                SeriesSource::Timeseries(_) => "#c00eae",
                SeriesSource::Events => "#4c7aff",
                SeriesSource::TargetLatency => "#63eaad",
            },
            tick_color: "#a4a4a4".to_owned(),
            x_formatter: FormatterKind::Time,
            y_formatter: FormatterKind::Duration,
        },
    );

    let svg_lines = svg
        .replace('>', ">\n")
        .split('\n')
        .filter(|line| !line.is_empty())
        .map(str::to_owned)
        .collect::<Vec<_>>();
    insta::assert_debug_snapshot!(svg_lines);
}

fn get_metric_at_minute(minute: u8, value: f64) -> Metric {
    Metric::builder()
        .time(get_date_at_minute(minute))
        .value(value)
        .build()
}

fn get_date_at_minute(minute: u8) -> Timestamp {
    Timestamp::from_str(&format!("2023-07-18T16:{minute:02}:00.000Z")).unwrap()
}

fn get_event_at_minute(minute: u8, title: &str) -> ProviderEvent {
    ProviderEvent::builder()
        .time(get_date_at_minute(minute))
        .title(title)
        .build()
}
