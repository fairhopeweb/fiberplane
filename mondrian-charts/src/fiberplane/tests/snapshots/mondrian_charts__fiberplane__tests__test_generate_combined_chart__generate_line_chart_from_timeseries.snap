---
source: mondrian-charts/src/fiberplane/tests/test_generate_combined_chart.rs
expression: chart
---
xAxis:
  minValue: 1689696000
  maxValue: 1689696600
  tickSuggestions:
    - 1689696000
    - 1689696060
    - 1689696120
    - 1689696180
    - 1689696240
    - 1689696300
    - 1689696360
    - 1689696420
    - 1689696480
    - 1689696540
yAxis:
  minValue: 9
  maxValue: 40
  tickSuggestions: ~
shapeLists:
  - shapes:
      - type: line
        points:
          - x: 0
            y: 0.03225806451612903
            source:
              type: metric
              time: "2023-07-18T16:00:00Z"
              attributes: {}
              resource: {}
              value: 10
          - x: 0.1
            y: 0.1935483870967742
            source:
              type: metric
              time: "2023-07-18T16:01:00Z"
              attributes: {}
              resource: {}
              value: 15
          - x: 0.2
            y: 0.3548387096774194
            source:
              type: metric
              time: "2023-07-18T16:02:00Z"
              attributes: {}
              resource: {}
              value: 20
      - type: line
        points:
          - x: 0.4
            y: 0.5161290322580645
            source:
              type: metric
              time: "2023-07-18T16:04:00Z"
              attributes: {}
              resource: {}
              value: 25
          - x: 0.5
            y: 0.6774193548387096
            source:
              type: metric
              time: "2023-07-18T16:05:00Z"
              attributes: {}
              resource: {}
              value: 30
          - x: 0.6
            y: 0.5161290322580645
            source:
              type: metric
              time: "2023-07-18T16:06:00Z"
              attributes: {}
              resource: {}
              value: 25
          - x: 0.7
            y: 0.3548387096774194
            source:
              type: metric
              time: "2023-07-18T16:07:00Z"
              attributes: {}
              resource: {}
              value: 20
          - x: 0.8
            y: 0.1935483870967742
            source:
              type: metric
              time: "2023-07-18T16:08:00Z"
              attributes: {}
              resource: {}
              value: 15
          - x: 0.9
            y: 0.03225806451612903
            source:
              type: metric
              time: "2023-07-18T16:09:00Z"
              attributes: {}
              resource: {}
              value: 10
    source:
      type: timeseries
      name: dummy_data
      labels: {}
      metrics:
        - time: "2023-07-18T16:00:00Z"
          attributes: {}
          resource: {}
          value: 10
        - time: "2023-07-18T16:01:00Z"
          attributes: {}
          resource: {}
          value: 15
        - time: "2023-07-18T16:02:00Z"
          attributes: {}
          resource: {}
          value: 20
        - time: "2023-07-18T16:04:00Z"
          attributes: {}
          resource: {}
          value: 25
        - time: "2023-07-18T16:05:00Z"
          attributes: {}
          resource: {}
          value: 30
        - time: "2023-07-18T16:06:00Z"
          attributes: {}
          resource: {}
          value: 25
        - time: "2023-07-18T16:07:00Z"
          attributes: {}
          resource: {}
          value: 20
        - time: "2023-07-18T16:08:00Z"
          attributes: {}
          resource: {}
          value: 15
        - time: "2023-07-18T16:09:00Z"
          attributes: {}
          resource: {}
          value: 10
      attributes: {}
      resource: {}
      visible: true
  - shapes:
      - type: area
        points:
          - x: 0.2
            yMin: 0
            yMax: 1
            source:
              type: event
              time: "2023-07-18T16:02:00Z"
              attributes: {}
              resource: {}
              title: deploy 1
              labels: {}
      - type: area
        points:
          - x: 0.7
            yMin: 0
            yMax: 1
            source:
              type: event
              time: "2023-07-18T16:07:00Z"
              attributes: {}
              resource: {}
              title: deploy 2
              labels: {}
    source:
      type: events
  - shapes:
      - type: line
        points:
          - x: 0
            y: 1
            source:
              type: none
          - x: 1
            y: 1
            source:
              type: none
    source:
      type: target_latency

